pkgs/applications/misc/nb/default.nix

```diff
@@ -0,0 +1,71 @@
+{ stdenv, lib, fetchFromGitHub, installShellFiles
+, bash, git
+}:
+stdenv.mkDerivation rec {
+  version = "5.7.8";
+  pname = "nb";
+
+  src = fetchFromGitHub {
+    owner = "xwmx";
+    repo = "nb";
+    rev = version;
+    sha256 = "1sw09sgskzsn3dy4ipvl52hwh5k3cwqfhjnm7rmxbzcpfk3jm3sp";
+  };
+
+  buildInputs = [ bash git ];
+  nativeBuildInputs = [ installShellFiles ];
+
+  dontBuild = true;
+
+  installPhase = ''
+    mkdir -p $out/bin/
+    mv nb $out/bin/
+    mv bin/bookmark $out/bin/
+    
+    runHook postInstall
+  '';
+
+  postInstall = ''
+    installShellCompletion etc/nb-completion.{bash,zsh}
+  '';
+
+  meta = with lib; {
+    description = "A command line note-taking, bookmarking, archiving, and knowledge base application";
+    longDescription = ''
+      `nb` creates notes in text-based formats like Markdown, Emacs Org mode,
+      and LaTeX, can work with files in any format, can import and export notes
+      to many document formats, and can create private, password-protected
+      encrypted notes and bookmarks. With `nb`, you can write notes using Vim,
+      Emacs, VS Code, Sublime Text, and any other text editor you like. `nb`
+      works in any standard Linux / Unix environment, including macOS and
+      Windows via WSL. Optional dependencies can be installed to enhance
+      functionality, but `nb` works great without them.
+
+      `nb` is also a powerful text-based CLI bookmarking system. Page
+      information is automatically downloaded, compiled, and saved into normal
+      Markdown documents made for humans, so bookmarks are easy to edit just
+      like any other note.
+
+      `nb` uses Git in the background to automatically record changes and sync
+      notebooks with remote repositories. `nb` can also be configured to sync
+      notebooks using a general purpose syncing utility like Dropbox so notes
+      can be edited in other apps on any device.
+
+      `nb` is designed to be portable, future-focused, and vendor independent,
+      providing a full-featured and intuitive experience within a highly
+      composable user-centric text interface. The entire program is a single
+      well-tested shell script that can be installed, copied, or curled almost
+      anywhere and just work, using progressive enhancement for various
+      experience improvements in more capable environments. `nb` works great
+      whether you have one notebook with just a few notes or dozens of
+      notebooks containing thousands of notes, bookmarks, and other items. `nb`
+      makes it easy to incorporate other tools, writing apps, and workflows.
+      `nb` can be used a little, a lot, once in a while, or for just a subset
+      of features. `nb` is flexible.
+    '';
+    homepage = "https://xwmx.github.io/nb/";
+    license = licenses.agpl3Plus;
+    maintainers = [ maintainers.toonn ];
+    platforms = platforms.all;
+  };
+}
```

Comment:
```suggestion
{ stdenv
, lib
, fetchFromGitHub
, makeWrapper
, installShellFiles
, bash
, git
# Optional dependencies
, bat
, nmap
, pandoc
, ripgrep
, tig
, w3m
, ack
, silver-searcher
, catimg
, exa
#, arcan
, imagemagick
, gnupg
, highlight
, imgcat
, links2
, lynx
, mc
, mpg123
, mplayer
#, xpdf
, python310Packages
, ranger
#, nodePackages
, ripgrep-all
, termpdfpy
, vifm
}:

stdenv.mkDerivation rec {
  version = "6.10.2";
  pname = "nb";

  src = fetchFromGitHub {
    owner = "xwmx";
    repo = "nb";
    rev = version;
    sha256 = "sha256-n87NT+qzBq8J1n0aT2Di7iBEN9inDWXAb8c7Qd2y0TA=";
  };

  nativeBuildInputs = [ makeWrapper installShellFiles ];

  dontBuild = true;

  postPatch = ''
    patchShebangs ./nb ./bin/bookmark
  '';

  installPhase = ''
    mkdir -p $out/bin
    mv nb $out/bin/
    mv bin/bookmark $out/bin/
    wrapProgram $out/bin/nb --prefix PATH : "${lib.makeBinPath [
      bash
      git
      bat
      nmap
      pandoc
      ripgrep
      tig
      w3m
      ack
      silver-searcher
      catimg
      exa
      #arcan.ffmpeg
      imagemagick
      gnupg
      highlight
      imgcat
      links2
      lynx
      mc
      mpg123
      mplayer
      # Insecure
      #xpdf
      python310Packages.pygments
      ranger
      # Build issues
      #nodePackages.readability-cli
      ripgrep-all
      termpdfpy
      vifm
    ]}"
    
    runHook postInstall
  '';

  postInstall = ''
    installShellCompletion etc/nb-completion.{bash,zsh}
  '';

  meta = with lib; {
    description = "A command line note-taking, bookmarking, archiving, and knowledge base application";
    longDescription = ''
      `nb` creates notes in text-based formats like Markdown, Emacs Org mode,
      and LaTeX, can work with files in any format, can import and export notes
      to many document formats, and can create private, password-protected
      encrypted notes and bookmarks. With `nb`, you can write notes using Vim,
      Emacs, VS Code, Sublime Text, and any other text editor you like. `nb`
      works in any standard Linux / Unix environment, including macOS and
      Windows via WSL. Optional dependencies can be installed to enhance
      functionality, but `nb` works great without them.

      `nb` is also a powerful text-based CLI bookmarking system. Page
      information is automatically downloaded, compiled, and saved into normal
      Markdown documents made for humans, so bookmarks are easy to edit just
      like any other note.

      `nb` uses Git in the background to automatically record changes and sync
      notebooks with remote repositories. `nb` can also be configured to sync
      notebooks using a general purpose syncing utility like Dropbox so notes
      can be edited in other apps on any device.

      `nb` is designed to be portable, future-focused, and vendor independent,
      providing a full-featured and intuitive experience within a highly
      composable user-centric text interface. The entire program is a single
      well-tested shell script that can be installed, copied, or curled almost
      anywhere and just work, using progressive enhancement for various
      experience improvements in more capable environments. `nb` works great
      whether you have one notebook with just a few notes or dozens of
      notebooks containing thousands of notes, bookmarks, and other items. `nb`
      makes it easy to incorporate other tools, writing apps, and workflows.
      `nb` can be used a little, a lot, once in a while, or for just a subset
      of features. `nb` is flexible.
    '';
    homepage = "https://xwmx.github.io/nb/";
    license = licenses.agpl3Plus;
    maintainers = with maintainers; [ toonn ];
    platforms = platforms.all;
  };
}
```

Reply:
I tried to make the wrapProgram function work for both nb and bookmark with a for statement but that didn't work. Is there another way than just cloning it for bookmark?

Reply:
Also, should be optional dependencies be included here or not?

Reply:
Do we need to wrap it separately? I think you can invoke it as a subcommand and it's a pretty generic name to claim.

I'm not really sure about the optional deps. They were atrociously badly documented when I last looked into this, most weren't actually mentioned anywhere in the docs. They're not build-time dependencies and `nb` simply picks them up from the environment if they're present so I don't think we should burden ourselves maintaining a potentially volatile list of deps.

Reply:
I removed the optional deps.